File: Vbf/py/vbf.py
#!/usr/bin/env python3
# Original author: Felice Pantaleo (CERN) <felice.pantaleo@cern.ch>

"""What are the dijet mass and the rapidity gap of the two quarks that recoil against the Higgs?

    python3 vbf.py ../../Common/fixtures/vbf.json
    python3 vbf.py step3.root -n 2

The C++ twin is cpp/Vbf.cc and prints the same lines.
"""

import argparse
import sys

from TruthGraphAnalysis.Common.exampleSupport import add, eta, eventNumber, graphsFrom, isParton, mass, pt


def run(graph, out=sys.stdout):
    """Prints one line per object of interest."""
    higgs = graph.signalParticles()
    tagging = [s for s in graph.productionSiblings(higgs[0]) if isParton(graph.pdgId(s))] if higgs else []
    if len(tagging) < 2:
        out.write("vbf: %d Higgs, %d tagging partons\n" % (len(higgs), len(tagging)))
        return
    tagging.sort(key=lambda i: -pt(graph.p4(i)))
    j1, j2 = tagging[0], tagging[1]
    out.write("vbf: %d Higgs, tagging partons %d %d: m(jj) %.2f GeV, |delta eta| %.2f\n"
              % (len(higgs), graph.pdgId(j1), graph.pdgId(j2), mass(add(graph.p4(j1), graph.p4(j2))),
                 abs(eta(graph.p4(j1)) - eta(graph.p4(j2)))))


def main():
    parser = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument("-n", "--maxEvents", type=int, default=-1)
    parser.add_argument("inputs", nargs="+", help="an EDM file, or one or more JSON dumps")
    args = parser.parse_args()
    for graph in graphsFrom(args.inputs, args.maxEvents):
        print("== event %s" % eventNumber(graph))
        run(graph)


if __name__ == "__main__":
    main()

File: Vbf/cpp/Vbf.cc
// Original author: Felice Pantaleo (CERN) <felice.pantaleo@cern.ch>

#include "TruthGraphAnalysis/Vbf/cpp/Vbf.h"

#include <cmath>
#include <string>
#include <vector>

#include "PhysicsTools/TruthInfo/interface/Branch.h"
#include "PhysicsTools/TruthInfo/interface/TruthLevels.h"
#include "TruthGraphAnalysis/Common/interface/ExampleSupport.h"

namespace tga::vbf {

  using tga::fixed;

  void run(truth::Graph const& graph, std::ostream& out) {
    const auto higgs = graph.signalParticles();
    // The tagging quarks are the partons produced with the Higgs. The parton-jet level is
    // not the answer: it also holds the quarks a hadronic Higgs decay makes.
    std::vector<truth::Particle> tagging;
    if (!higgs.empty()) {
      for (auto const& sibling : higgs.front().productionSiblings()) {
        if (truth::isParton(sibling.pdgId())) {
          tagging.push_back(sibling);
        }
      }
    }
    if (tagging.size() < 2) {
      out << "vbf: " << higgs.size() << " Higgs, " << tagging.size() << " tagging partons\n";
      return;
    }
    std::stable_sort(tagging.begin(), tagging.end(), [](auto const& a, auto const& b) {
      return a.momentum().pt() > b.momentum().pt();
    });
    const auto dijet = tagging[0].momentum() + tagging[1].momentum();
    out << "vbf: " << higgs.size() << " Higgs, tagging partons " << tagging[0].pdgId() << " " << tagging[1].pdgId()
        << ": m(jj) " << fixed(dijet.mass()) << " GeV, |delta eta| "
        << fixed(std::abs(tagging[0].momentum().eta() - tagging[1].momentum().eta())) << "\n";
  }

}  // namespace tga::vbf
